TST: add contract for filters

ambrosejcarr opened the pull request:
This PR adds a test that acts as a contract for filters that extend `FilterAlgorithmBase`

See [added test](https://github.com/spacetx/starfish/compare/ajc-floats-only...ajc-filter-testing-contract?expand=1#diff-2cf2138a0682ad915b3f694085559a94R1) for contract contents. 

Of note, imposes that Filters always return an ImageStack, even if run `in_place=True` to close #457.

Some tests needed to be adapted to adhere to the tests: 
- ScaleByPercentile has an interesting interaction with float [0, 1] range wherein it very often goes out of range. I used the util function to make sure it stays inside, but @dganguli should address whether this still works as intended. See change [here](https://github.com/spacetx/starfish/compare/ajc-floats-only...ajc-filter-testing-contract?expand=1#diff-7c759799493be844aca9aceb7752edcbR59)
- Added multiprocessing to ZeroByChannelMagnitude to make it accept `n_processes`. It does something a bit different, so can't use apply directly.

ambrosejcarr commented:
Correct. Rebased.

ttung commented on starfish/image/_filter/scale_by_percentile.py:
```diff
@@ -54,16 +55,15 @@ def scale(image: np.ndarray, p: int) -> np.ndarray:
-        dtype = image.dtype
         image = image / v
-        return image.astype(dtype)
+        image = preserve_float_range(image)
```
why is this necessary?

ambrosejcarr replied:
if you scale by anything other than the 100th percentile (excluding ties), it will generate values above 1: 

e.g. if the top value is `0.96` and you scale by the 90th percentile, which for this example say is `0.87`, you'll get `1.10`.

ttung commented on starfish/image/_filter/zero_by_channel_magnitude.py:
```diff
@@ -22,21 +25,50 @@ def __init__(self, thresh: int, normalize: bool, **kwargs) -> None:
         normalize : bool
             if True, this scales all rounds to have unit L2 norm across channels
         """
+        if is_volume is True:
```
can just be:

`if is_volume:`

ttung commented on starfish/image/_filter/zero_by_channel_magnitude.py:
```diff
@@ -22,21 +25,50 @@ def __init__(self, thresh: int, normalize: bool, **kwargs) -> None:
+            '--thresh', type=int, help="minimum magnitude threshold for pixels across channels")
+
+    @staticmethod
+    def zero_by_channel_magnitude(r_dat: Tuple[int, xr.DataArray], thresh, stack, normalize):
```
pls add missing type hints here.

ttung commented on starfish/test/image/filter/test_api_contract.py:
```diff
@@ -0,0 +1,89 @@
+"""
+This module stores a contract for new Filter algorithms. To add a new filter, simply add default
```
It would be useful to know what this contract is.

ttung commented on starfish/test/image/filter/test_api_contract.py:
```diff
@@ -0,0 +1,89 @@
+
+    # output is dtype float and within the expected interval of [0, 1]
+    assert filtered.numpy_array.dtype == np.float32, f'{filter_class} must output float32 data'
+    assert np.all(filtered.numpy_array >= 0), \
```
Should we make this assertion?  I think @dganguli felt otherwise.....?

ambrosejcarr replied:
So far, we're requiring this be the case. If we make changes in later PRs, this can be removed.

ttung commented on starfish/image/_filter/bandpass.py:
```diff
@@ -83,7 +83,7 @@ def bandpass(
     def run(
             self, stack: ImageStack, in_place: bool=True, verbose: bool=False,
             n_processes: Optional[int]=None
-    ) -> Optional[ImageStack]:
```
Making filters always return an image could have been a separate EZPZ PR.

ambrosejcarr commented:
I've removed the multiprocessing from the scale by channel magnitude filter and rebased to fetch changes to PipelineComponent. All tests are passing now, and I think this should be good to merge.

ambrosejcarr commented:
Note: I added a line to requirements.txt to grab `pycodestyle<2.4.0`, as that version breaks flake8 due to the use of a wildcard character in our flake8 config.

ttung commented:
hm... where do we have wildcard characters in our flake8 config?

```
[flake8]
max-line-length=100
ignore: E301, E302, E305, E731, F811
application-import-names = starfish, examples, validate_sptx
import-order-style = smarkets
```

ambrosejcarr commented:
Sorry, it's in the Makefile: 
```make
fast:	lint mypy test

lint:   lint-non-init lint-init

lint-non-init:
	flake8 --ignore 'E301, E302, E305, E401, E731, F811' --exclude='*__init__.py' $(MODULES)

lint-init:
	flake8 --ignore 'E301, E302, E305, E401, F401, E731, F811' --filename='*__init__.py' $(MODULES)
```

```bash
$ make lint
flake8 --ignore 'E301, E302, E305, E401, E731, F811' --exclude='*__init__.py' starfish examples validate_sptx
Traceback (most recent call last):
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2333, in resolve
    return functools.reduce(getattr, self.attrs, module)
AttributeError: module 'pycodestyle' has no attribute 'break_around_binary_operator'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 178, in load_plugin
    self._load(verify_requirements)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 150, in _load
    self._plugin = resolve()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2335, in resolve
    raise ImportError(str(exc))
ImportError: module 'pycodestyle' has no attribute 'break_around_binary_operator'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ajc/projects/spacetx/starfish/.venv/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/cli.py", line 16, in main
    app.run(argv)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 396, in run
    self._run(argv)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 383, in _run
    self.initialize(argv)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 365, in initialize
    self.find_plugins()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 191, in find_plugins
    self.check_plugins.load_plugins()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 423, in load_plugins
    plugins = list(self.manager.map(load_plugin))
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 311, in map
    yield func(self.plugins[name], *args, **kwargs)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 421, in load_plugin
    return plugin.load_plugin()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 186, in load_plugin
    raise failed_to_load
flake8.exceptions.FailedToLoadPlugin: Flake8 failed to load plugin "pycodestyle.break_around_binary_operator" due to module 'pycodestyle' has no attribute 'break_around_binary_operator'.
make: *** [lint-non-init] Error 1
```

ambrosejcarr commented:
To reproduce, just `pip install --upgrade pycodestyle` and run `make lint`

ttung commented on starfish/test/image/filter/test_api_contract.py:
```diff
@@ -0,0 +1,106 @@
+# store some default parameters for testing purposes. New algorithms will have to add parameters to
+# this in order to not fail tests
+default_filtering_parameters = {
+    'Bandpass': dict(lshort=1, llong=3, threshold=0.01),
```
Does it make sense to put these defaults in the algorithms classes?

ambrosejcarr replied:
Potentially as a private constant `_TESTING_DEFAULTS`. Definitely not exposed to users. I like the former, what do you think?

ambrosejcarr commented on starfish/image/_filter/zero_by_channel_magnitude.py:
```diff
@@ -50,6 +52,8 @@ def run(
         verbose : bool
             if True, report on the percentage completed during processing (default = False)
+        n_processes : Optional[int]: None
+            Not implemented. Number of processes to use when applying filter.
```
Fast enough that it's not a concern, I think.

ttung commented on starfish/test/image/filter/test_api_contract.py:
```diff
@@ -0,0 +1,91 @@
+def test_all_methods_adhere_to_contract(filter_class):
+    """Test that all filter algorithms adhere to the filtering contract"""
+
+    default_kwargs = filter_class._DEFAULT_TESTING_PARAMETERS
```
necro comment: `FilterAlgorithmBase` should be documented to reflect that this magic classvar is a requirement.
